vault/identity_store_oidc_provider.go

```diff
@@ -161,3 +216,129 @@ func (i *IdentityStore) pathOIDCAssignmentExistenceCheck(ctx context.Context, re
+				return nil, err
+			}
+		}
+	}
+
+	if descriptionRaw, ok := d.GetOk("description"); ok {
+		scope.Description = descriptionRaw.(string)
+	} else if req.Operation == logical.CreateOperation {
+		scope.Description = d.GetDefaultOrZero("description").(string)
+	}
+
+	if templateRaw, ok := d.GetOk("template"); ok {
+		scope.Template = templateRaw.(string)
+	} else if req.Operation == logical.CreateOperation {
+		scope.Template = d.GetDefaultOrZero("template").(string)
+	}
+
+	// Attempt to decode as base64 and use that if it works
+	if decoded, err := base64.StdEncoding.DecodeString(scope.Template); err == nil {
+		scope.Template = string(decoded)
+	}
+
+	// Validate that template can be parsed and results in valid JSON
+	if scope.Template != "" {
+		_, populatedTemplate, err := identitytpl.PopulateString(identitytpl.PopulateStringInput{
+			Mode:   identitytpl.JSONTemplating,
+			String: scope.Template,
+			Entity: new(logical.Entity),
+			Groups: make([]*logical.Group, 0),
+			// namespace?
+		})
+		if err != nil {
+			return logical.ErrorResponse("error parsing template: %s", err.Error()), nil
+		}
+
+		var tmp map[string]interface{}
+		if err := json.Unmarshal([]byte(populatedTemplate), &tmp); err != nil {
+			return logical.ErrorResponse("error parsing template JSON: %s", err.Error()), nil
+		}
+
+		for key := range tmp {
+			if strutil.StrListContains(requiredClaims, key) {
+				return logical.ErrorResponse(`top level key %q not allowed. Restricted keys: %s`,
+					key, strings.Join(requiredClaims, ", ")), nil
+			}
+		}
+	}
+	// store scope
+	entry, err := logical.StorageEntryJSON(scopePath+name, scope)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := req.Storage.Put(ctx, entry); err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+// pathOIDCListScope is used to list scopes
+func (i *IdentityStore) pathOIDCListScope(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	scopes, err := req.Storage.List(ctx, scopePath)
+	if err != nil {
+		return nil, err
+	}
+	return logical.ListResponse(scopes), nil
+}
+
+// pathOIDCReadScope is used to read an existing scope
+func (i *IdentityStore) pathOIDCReadScope(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	name := d.Get("name").(string)
+
+	entry, err := req.Storage.Get(ctx, scopePath+name)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, nil
+	}
+
+	var scope scope
+	if err := entry.DecodeJSON(&scope); err != nil {
+		return nil, err
+	}
+	return &logical.Response{
+		Data: map[string]interface{}{
+			"template":    scope.Template,
+			"description": scope.Description,
+		},
+	}, nil
+}
+
+// pathOIDCDeleteScope is used to delete an scope
+func (i *IdentityStore) pathOIDCDeleteScope(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	name := d.Get("name").(string)
+	err := req.Storage.Delete(ctx, scopePath+name)
```

Comment:
Leaving a note that we'll likely need to prevent deletion of scopes that are being referenced by providers. We can add that check with the provider PR (similar to the client/assignment situation). Won't block this PR 👍

Reply:
CC @vinay-gopalan

Reply:
Just realized that Vinay's PR is branched from the `oidc-provider-assignment-api` branch which does not yet have the Scope code. Maybe this can be addressed in a separate issue after all is merged?

Reply:
If need be I can merge the this branch into the `oidc-provider-api` branch to get access to the Scope code. Things should be sorted as long as we merge the assignment and scope API PRs before the Provider API, I think